### PR TITLE
[ROO-70] [mlir] [arith] emulate wide int

### PR DESCRIPTION
**Pull request by egebeysel**
Well, emulation pass for `arith.fptoui` and `arith.fptosi`.

The basic algorithm looks like this:

```c
const double TWO_POW_BW = (uint_64_t(1) << bitwidth); // 2^BW

// f is a floating-point value representing the 64-bit number.
uint32_t hi = (uint32_t)(f / TWO_POW_BW);         // Truncates the division result.
uint32_t lo = (uint32_t)(f - hi * TWO_POW_BW);       // Subtracts to get the lower bits.
```

`f - hi * TWO_POW_BW` is emitted via `arith.remf`.

`arith.fptosi` emits `fptoui`  with the absolute value of the input fp. It also does a bounds check and emits `MAX_SIGNED_INT` or `MIN_SIGNED_INT` w.r.t. the sign of the input.

I added the runner tests, but here are some remarks:

According to LLVM LangRef https://llvm.org/docs/LangRef.html#fptoui-to-instruction "If the value cannot fit in target int type, the result is a poison value.". So basically, the `+-inf` and overflow values - but they result in the same poison value in the case of `fptoui` when we run it with and without emulation (see the integration test). Also, `NaN`  values result in different results with `fptosi` and `fptoui`, namely `-2^63(INT64_MIN)` and `-1` respectively. I'm not entirely sure if this is UB/poison or not.

Lastly, numbers that are representable with unsigned integers but not with signed ones (`>=2^63` in the case of `int64`) also result in poison-looking numbers: `fptosi(2^63)` emits `-2^63` with the `mlir-cpu-runner` without the emulation, which seems vague to be honest.

https://llvm.org/docs/LangRef.html#behavior-of-floating-point-nan-values also does not say much about the behavior of NaN bitcast/conversion ops. 

https://llvm.org/docs/LangRef.html#llvm-fptoui-sat-intrinsic gives saturating semantics but default lowering does not result in these. I guess not doing anything and leaving it to the specific target/lowering might emit these? Not sure...

So if I actually have to handle the cases of `abs(fp) >= MAX_SINT` in `arith.fptosi` and overflows in `arith.fptosi` are unclear. I guess the best course of action here would be that I post the PR upstream and ask for reviews over there, after your rounds of review.

### CODE REVIEW

**Comment by egebeysel** [comment]
* **#3** <a href="https://app.graphite.dev/github/pr/RooflineAI/llvm-project/3?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/RooflineAI/llvm-project/3?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `integrate-llvm-project-20250217`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Comment by maxbartel** [comment]
Can you create LIT tests? Will be a lot easier to review with them

**Comment by egebeysel** [comment]
yes, sorry, was going to do it the next time I'm working

**Comment by egebeysel** [comment]
> Can you create LIT tests? Will be a lot easier to review with them

I just pushed the LIT Tests for the `fptoui` in scalar and vector forms and `fptosi` in scalar form. I'll also push the vector form of the latter when I have time. Also, I tried keeping the tests as _relevant_ to the actual operation and isolated from other _emulations_ as possible, but am open to any suggestions over here.

**Comment by egebeysel** [comment]
I hope this makes it a little bit easier to review, although the tests themselves are also pretty crowded.

**Comment by egebeysel** [comment]
Also, please write up the naming suggestions for the variables in code/tests if they're really not understandable :)

File: mlir/test/Dialect/Arith/emulate-wide-int.mlir [hunk 1]
```diff
@@ -1007,3 +1007,152 @@ func.func @sitofp_i64_f64_vector(%a : vector<3xi64>) -> vector<3xf64> {
     %r = arith.sitofp %a : vector<3xi64> to vector<3xf64>
     return %r : vector<3xf64>
 }
+
```
Comment (chrsmcgrr) [inline mlir/test/Dialect/Arith/emulate-wide-int.mlir:1010]: I would like to see some CPU runner tests as well. Looking at the original PR https://reviews.llvm.org/D146606 for the other conversion direction you can add some test cases there. Just to check simple values and edge cases.

**Review by chrsmcgrr — CHANGES_REQUESTED** [review]
Ok I'm going to try and review the code this week, but just want to submit my comment for also adding runner tests to make sure we are functionally correct at least with a simple CPU pipeline.

**Comment by egebeysel** [comment]
added runner tests 👍

File: mlir/test/Integration/Dialect/Arith/CPU/test-wide-int-emulation-fptoui-i64.mlir [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+  // CHECK-NEXT:    4294967295
+  func.call @check_fptoui(%cst_i32_max) : (f64) -> ()
+  // CHECK-NEXT:    4294967296
+  func.call @check_fptoui(%cst_i32_overflow) : (f64) -> ()
+  // CHECK-NEXT:    1099511627776
+  func.call @check_fptoui(%cstpow40) : (f64) -> ()
+  // CHECK-NEXT:    1099512676352
+  func.call @check_fptoui(%cst_pow40ppow20) : (f64) -> ()
+  // CHECK-NEXT:    -1
+  func.call @check_fptoui(%cst_i64_overflow) : (f64) -> ()
```
Comment (egebeysel) [inline mlir/test/Integration/Dialect/Arith/CPU/test-wide-int-emulation-fptoui-i64.mlir]: So I'm not sure if I'm supposed to handle/check the cases below this

File: mlir/test/Integration/Dialect/Arith/CPU/test-wide-int-emulation-fptosi-i64.mlir [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  func.call @check_fptosi(%cst_i32_noverflow) : (f64) -> ()
+  // CHECK-NEXT:    1099511627776
+  func.call @check_fptosi(%cstpow40) : (f64) -> ()
+  // CHECK-NEXT:    -1099511627776
+  func.call @check_fptosi(%cstnpow40) : (f64) -> ()
+  // CHECK-NEXT:    1099512676352
+  func.call @check_fptosi(%cst_pow40ppow20) : (f64) -> ()
+  // CHECK-NEXT:    -1099512676352
+  func.call @check_fptosi(%cst_npow40ppow20) : (f64) -> ()
+
```
Comment (egebeysel) [inline mlir/test/Integration/Dialect/Arith/CPU/test-wide-int-emulation-fptosi-i64.mlir:80]: Not sure to handle the nan, +-inf and overflow cases over here or to handle them in the emulation, see description

File: mlir/test/Integration/Dialect/Arith/CPU/test-wide-int-emulation-fptoui-i64.mlir (hunk 2 above)
Comment (chrsmcgrr) [reply mlir/test/Integration/Dialect/Arith/CPU/test-wide-int-emulation-fptoui-i64.mlir]: So it seems that for float->int IEEE defines it as this:
```
If a NaN, infinite, or out-of-range numeric value cannot be represented in the destination format and no other indication is possible, an invalid operation exception will be triggered.
```
Which is what for example we see with x86: https://www.felixcloutier.com/x86/vcvtss2usi. So its left to the platform to handle this it seems, not the language AFAICS. 

So I would assume that all the arith ops we use for this pattern should also handle these special cases and it is left to the platform to signal an exception. What I am surprised about is we get an actual value out. 

Since we are using a lower-bitwdith `fptoui` anyway I don't think this transform can do anything to handle these edge cases. Maybe only ensure that if there is a overflow we set both high and low to one of these special values.

@maxbartel what do you think?

File: mlir/test/Dialect/Arith/emulate-wide-int.mlir [hunk 4]
```diff
@@ -1007,3 +1007,173 @@ func.func @sitofp_i64_f64_vector(%a : vector<3xi64>) -> vector<3xf64> {
...
+func.func @fptoui_i64_f64_vector(%a : vector<3xf64>) -> vector<3xi64> {
+    %r = arith.fptoui %a : vector<3xf64> to vector<3xi64>
+    return %r : vector<3xi64>
+}
+
+// This generates lines that are already verified by other patterns
+// We do not re-verify these and just check for the wrapper around fptoui by following its low part
+// CHECK-LABEL:     func @fptosi_i64_f64
+// CHECK-SAME:      ([[ARG:%.+]]: f64) -> vector<2xi32>
+// CHECK:           [[MAXSINT:%.+]] = arith.constant 9.2233720368547758E+18 : f64
```
Comment (chrsmcgrr) [inline mlir/test/Dialect/Arith/emulate-wide-int.mlir]: The max i64 integer is given as `9223372036854775807` does the string printing cutoff here?

File: mlir/lib/Dialect/Arith/Transforms/EmulateWideInt.cpp [hunk 5]
```diff
@@ -974,6 +974,154 @@ struct ConvertUIToFP final : OpConversionPattern<arith::UIToFPOp> {
...
+    Value remainder =
+        rewriter.create<arith::RemFOp>(loc, inFp, powBitwidthFloatCst);
+    Value resLow =
+        rewriter.create<arith::FPToUIOp>(loc, newHalfType, remainder);
+    // If we have the overflow case, emit the high part as the low part as well,
+    // since that already handles this case.
+    Value isOverflow = rewriter.create<arith::CmpFOp>(
+        loc, arith::CmpFPredicate::OGE, fpDivPowBitwidth, powBitwidthFloatCst);
+    Value lowerPart =
+        rewriter.create<arith::SelectOp>(loc, isOverflow, resHigh, resLow);
```
Comment (chrsmcgrr) [inline mlir/lib/Dialect/Arith/Transforms/EmulateWideInt.cpp]: I think we do not need to do this. It should be in theory ok to have:
```
<PoisonValue, LowValue>
```
as our high and low.

Because any math done should be with both parts in mind. The tricky thing is we cannot represent poison values in the integer format and we are relying on the lower precision fptoui to handle the signaling of a failure in the high part case.

**Review by chrsmcgrr — CHANGES_REQUESTED** [review]
I think its almost there. Just some changes maybe. And can you retarget this PR to our latest `integrate-iree-20250217` branch that would be great. 

Real tricky on what to do with the special cases. 

Just looking at what llvm does and gcc they seem to be more pragmatic in the emulation and just saturate when large or go to zero in other cases. 
https://github.com/llvm-mirror/compiler-rt/blob/master/lib/builtins/fp_fixuint_impl.inc#L16
https://gcc.gnu.org/onlinedocs/gccint/Soft-float-library-routines.html

Don't know if you read through these as well. AFAICS the special cases are undefined behavior.

File: mlir/test/Dialect/Arith/emulate-wide-int.mlir (hunk 4 above)
Comment (egebeysel) [reply mlir/test/Dialect/Arith/emulate-wide-int.mlir]: exactly, it does cutoff there

File: mlir/lib/Dialect/Arith/Transforms/EmulateWideInt.cpp [hunk 6]
```diff
@@ -974,6 +974,154 @@ struct ConvertUIToFP final : OpConversionPattern<arith::UIToFPOp> {
...
+    /* Negate the value if < 0 */
+    Value bitwiseNeg = rewriter.create<arith::XOrIOp>(loc, res, allOnesCst);
+    Value neg = rewriter.create<arith::AddIOp>(loc, bitwiseNeg, oneCst);
+    res = rewriter.create<arith::SelectOp>(loc, isNeg, neg, res);
+
+    /* Return the max/min signed int for greater values */
+    Value maxSIntCst =
+        createScalarOrSplatConstant(rewriter, loc, intTy, maxSInt);
+    Value minSIntCst =
+        createScalarOrSplatConstant(rewriter, loc, intTy, minSInt);
```
Comment (egebeysel) [inline mlir/lib/Dialect/Arith/Transforms/EmulateWideInt.cpp]: @chrsmcgrr last thing then,we don't need truncate like this and we just leave it to the platform then, right?

File: mlir/test/Dialect/Arith/emulate-wide-int.mlir (hunk 4 above)
Comment (chrsmcgrr) [reply mlir/test/Dialect/Arith/emulate-wide-int.mlir]: ok but I would expect that it would round for f64 using: `Integers between 2n and 2n+1 round to a multiple of 2n−52` for f64 which should be 9223372036854775808 AFAIK. Looks like an assembly print error? Can you check?

File: mlir/lib/Dialect/Arith/Transforms/EmulateWideInt.cpp (hunk 6 above)
Comment (chrsmcgrr) [reply mlir/lib/Dialect/Arith/Transforms/EmulateWideInt.cpp]: Yes, I think you are right. We should not be clamping to min and max. We will let the decomposition handle it.

File: mlir/test/Integration/Dialect/Arith/CPU/test-wide-int-emulation-fptosi-i64.mlir [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+  func.call @check_fptosi(%cstpow20) : (f64) -> ()
+  // CHECK-NEXT:    -1048576
+  func.call @check_fptosi(%cstnpow20) : (f64) -> ()
+  // CHECK-NEXT:    4294967295
+  func.call @check_fptosi(%cst_i32_max) : (f64) -> ()
+  // CHECK-NEXT:    -4294967296
+  func.call @check_fptosi(%cst_i32_min) : (f64) -> ()
+  // CHECK-NEXT:    4294967296
+  func.call @check_fptosi(%cst_i32_overflow) : (f64) -> ()
+    // CHECK-NEXT:    -4294967297
```
Comment (chrsmcgrr) [inline mlir/test/Integration/Dialect/Arith/CPU/test-wide-int-emulation-fptosi-i64.mlir]:
```suggestion
  // CHECK-NEXT:    -4294967297
```
nit formatting

File: mlir/test/Integration/Dialect/Arith/CPU/test-wide-int-emulation-fptoui-i64.mlir [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+   // CHECK-NEXT:    -1
+  func.call @check_fptoui(%cst_nan) : (f64) -> ()
```
Comment (chrsmcgrr) [inline mlir/test/Integration/Dialect/Arith/CPU/test-wide-int-emulation-fptoui-i64.mlir]: So running with godbolt and a c program we get a different value out then what we get here. And the assembly does use the x86 instruction (cvttsd2si) we expect. Can you check why?

see: https://godbolt.org/z/nK3n5Mcd9

**Review by chrsmcgrr — CHANGES_REQUESTED** [review]
In general it looks good to me. The LLVM lowering in CPU runner seems to generate different results than what we get with clang. I would be curious to know why. We just need to make sure its not this transform but another pass.

**Comment by egebeysel** [comment]
Do you mean with an emulated fptou/si cast or a non-emulated one? And different results for the edge cases or the normal ones? @chrsmcgrr

**Comment by chrsmcgrr** [comment]
I would expect the emulated fptoui should generate the same output as the non-emulated for `nan` and `+inf`/`-inf` what I would be curious to know is why it does not. The LIT CPU Runner produces `-1` in the case of `nan` and if I compile with clang on a x86_64 system I get `-9223372036854775808`.

**Comment by egebeysel** [comment]
For `nan` and `+-inf` I get `-1` with the emulated and non-emulated cases on the LIT CPU runner for the `fptoui` case. For the `fptosi` case, the non-emulated version emits `9223372036854775808` for all 3 while I get different results for these with the emulated case. But that was exactly what I was asking previously, if we have UB here anyways, how does that matter what result we emit over here? I don't think we should match the UB over here, do we? If we should, then I have to handle these as special cases of some sort, because I believe the _working on the absolute value_ is what's changing the result, so those would not be deferred to `fptoui` from the signed case.

Regarding clang producing different results, did you compile the result of the `mlir-opt` or did you just compile an equivalent C++ code? Can that be the case that you looked at the signed case with the clang program and the unsigned case with the LIT CPU runner?

@chrsmcgrr

File: mlir/test/Dialect/Arith/emulate-wide-int.mlir (hunk 4 above)
Comment (egebeysel) [reply mlir/test/Dialect/Arith/emulate-wide-int.mlir]: not doing the truncation anymore and the value is not emitted

File: mlir/test/Integration/Dialect/Arith/CPU/test-wide-int-emulation-fptoui-i64.mlir (hunk 8 above)
Comment (egebeysel) [reply mlir/test/Integration/Dialect/Arith/CPU/test-wide-int-emulation-fptoui-i64.mlir]: This one which converts the `nan` to `uint64_t` also emits the same cvttsd2si op and actually emits the same value whereas the same conversion with the LIT CPU Runner emit -1 in the unsigned case.

**Comment by egebeysel** [comment]
@chrsmcgrr looked into this a little bit more, the `fptosi` case yielding different results for `nan` and `-+inf` are related to the pass itself, but I don't really see a way of achieving that without special-casing for those cases, even if we don't defer to `fptoui` and handle the both parts in the signed case explicitly.

**Review by chrsmcgrr — APPROVED** [review]
Ok then let's merge it as is. At least we are fully aware of the corner cases. 

Make sure you rebase on the latest integration and then ping me here and I will merge this.

**Comment by egebeysel** [comment]
The latest integration is `integrate-llvm-project-20250217`, right? If so, I rebased it on that :) @chrsmcgrr

**Comment by chrsmcgrr** [comment]
@egebeysel exactly that's it. I will merge it now.
